File: scripts/sdk_smoke_arc_staging.py
"""Exercise the Python SDK against the FIXED V6 staging stack on Arc testnet.
Confirms client construction (arc-staging), credit_info, borrow, repay, loans
work against the 2026-08-fixed contracts. Uses the deployer account (already
onboarded + funded with staging MockUSDC via the JS smoke)."""
import os
import sys
import warnings

warnings.filterwarnings("ignore")
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "python"))

from web3 import Web3
from eth_account import Account
from specular.client import SpecularClient

RPC = os.environ.get("ARC_TESTNET_RPC_URL", "https://arc-testnet.drpc.org")

PASS = 0
FAIL = 0


def ok(label, cond, detail=""):
    global PASS, FAIL
    if cond:
        PASS += 1
        print(f"  PASS {label} {detail}")
    else:
        FAIL += 1
        print(f"  FAIL {label} {detail}")


def main():
    pk = os.environ.get("PRIVATE_KEY")
    if not pk:
        print("PRIVATE_KEY not set")
        sys.exit(1)
    w3 = Web3(Web3.HTTPProvider(RPC))
    acct = Account.from_key(pk)
    client = SpecularClient(w3, acct, network="arc-staging")
    ok("client targets fixed staging marketplace",
       client.marketplace_addr.lower() == "0xdbdf60ae5cb46d23aa44c062a4943655a6820f31")

    # Already onboarded by the JS smoke; onboard() is idempotent.
    client.onboard()

    info = client.credit_info()
    print(f"     score={info.score} limit={info.credit_limit_usdc} "
          f"collateral={info.collateral_pct}% rate={info.interest_rate_apr}%")
    ok("credit_info returns a score", isinstance(info.score, int))

    # Small collateralized borrow/repay cycle (fresh-tier agent posts collateral).
    res = client.borrow(2, 7)
    ok("borrow returns loanId", res.get("loanId") is not None, f"(loanId {res.get('loanId')})")
    tx = client.repay(res["loanId"])
    ok("repay returns tx hash", isinstance(tx, str) and len(tx) > 0)

    loans = client.loans()
    this = next((l for l in loans if l.loan_id == res["loanId"]), None)
    ok("loan shows REPAID", this is not None and this.state == "REPAID")

    print(f"\n=== Python SDK-vs-staging: {PASS} passed, {FAIL} failed ===")
    sys.exit(1 if FAIL else 0)


if __name__ == "__main__":
    main()

File: scripts/sdk-smoke-arc-staging.js
/**
 * Exercise the JS SDK (SpecularQuickstart) against the FIXED V6 staging stack on
 * Arc testnet. Confirms the SDK's onboard/creditInfo/supply/borrow/repay/claim/
 * loans all work against the 2026-08-fixed contracts with the launch levers ON.
 *
 * The staging agents are FRESH (score 0 → 100% collateral), unlike the old
 * arc-testnet e2e which assumes score-1000 agents — so this drives the
 * collateralized-borrow path (exercising the D2 exact-approval logic) and the
 * levers (M-1 creator-only borrow, F-C min supply, M-2 min-hold → no rep yet).
 *
 * Uses the deployer wallet (native USDC for gas + owner of the fresh MockUSDC,
 * so it can mint). Requires PRIVATE_KEY + ARC_TESTNET_RPC_URL.
 */
require('dotenv').config();
const { ethers } = require('ethers');
const fs = require('fs');
const path = require('path');
const { SpecularQuickstart } = require('../src/sdk/SpecularQuickstart.js');

const RPC = process.env.ARC_TESTNET_RPC_URL || 'https://arc-testnet.drpc.org';
const cfg = JSON.parse(fs.readFileSync(path.join(__dirname, '..', 'src', 'config', 'arc-testnet-v6-addresses.json'), 'utf8'));
const mockUsdcAbi = JSON.parse(fs.readFileSync(path.join(__dirname, '..', 'artifacts', 'contracts', 'tokens', 'MockUSDC.sol', 'MockUSDC.json'), 'utf8')).abi;

let pass = 0, fail = 0;
const ok = (label, cond, detail = '') => { if (cond) { pass++; console.log(`  ✅ ${label} ${detail}`); } else { fail++; console.log(`  ❌ ${label} ${detail}`); } };

async function main() {
    const provider = new ethers.JsonRpcProvider(RPC, 5042002, { batchMaxCount: 1 });
    const wallet = new ethers.Wallet(process.env.PRIVATE_KEY, provider);
    const sdk = new SpecularQuickstart(wallet, 'arc-staging');
    ok('SDK targets the fixed staging marketplace', sdk.addresses.marketplace.toLowerCase() === cfg.agentLiquidityMarketplace_v6.toLowerCase());

    // Ensure the deployer holds the staging MockUSDC (mint if needed).
    const usdc = new ethers.Contract(cfg.usdc, mockUsdcAbi, wallet);
    if ((await usdc.balanceOf(wallet.address)) < ethers.parseUnits('50', 6)) {
        await (await usdc.mint(wallet.address, ethers.parseUnits('1000', 6))).wait();
    }

    console.log('\n=== SDK.onboard() ===');
    const out = await sdk.onboard();
    ok('onboard returns agentId', out.agentId != null, `(agentId ${out.agentId})`);
    ok('onboard emits no blanket approval (D2 exact-approval)', out.approveTx === null);

    console.log('\n=== SDK.creditInfo() ===');
    const info = await sdk.creditInfo();
    console.log(`     score=${info.score} limit=${info.creditLimit} collateral=${info.collateralPct}% rate=${info.interestRateAPR}%`);
    ok('creditInfo returns a score', typeof info.score === 'number');
    // Fresh staging agent: score 0 → 100% collateral tier.
    ok('fresh agent is 100% collateral (levers/fresh state)', info.collateralPct === 100);

    console.log('\n=== SDK.supply() (F-C: >= 1 USDC) ===');
    await sdk.supply(out.agentId, 50); // supply 50 to own pool for borrow liquidity
    ok('supplied 50 to pool', true);

    console.log('\n=== SDK.borrow() + repay() (collateralized path) ===');
    const { loanId } = await sdk.borrow(5, 7); // 5 USDC, 7d — 100% collateral auto-approved by SDK
    ok('borrow returned loanId', loanId != null, `(loanId ${loanId})`);
    const repayHash = await sdk.repay(loanId);
    ok('repay returned tx hash', typeof repayHash === 'string' && repayHash.startsWith('0x'));

    console.log('\n=== SDK.loans() ===');
    const loans = await sdk.loans();
    const thisLoan = loans.find(l => l.id === Number(loanId));
    ok('loan shows REPAID', thisLoan && thisLoan.state === 'REPAID');

    console.log('\n=== SDK.claim() ===');
    const claimHash = await sdk.claim(out.agentId);
    ok('claim returned tx hash', typeof claimHash === 'string' && claimHash.startsWith('0x'));

    console.log(`\n=== SDK-vs-staging: ${pass} passed, ${fail} failed ===`);
    if (fail > 0) process.exit(1);
}

main().catch((e) => { console.error(e); process.exit(1); });
